### PR TITLE
fix several bugs

### PR DESCRIPTION
**Pull request by GiggleLiu**

### CODE REVIEW

**Comment by Roger-luo** [comment]
which bug exactly?

File: test/operations.jl [hunk 1]
```diff
@@ -22,4 +22,7 @@ end
     reg = rand_state(4)
     @test all(state(reg + (-reg)).==0)
     @test all(state(reg*2 - reg/0.5) .== 0)
+
+    reg = rand_state(3)
+    @test reg'*reg ≈ 1
```
Comment (GiggleLiu) [inline test/operations.jl]: Here is the fall back test for the bug. Which is mentioned in issue https://github.com/QuantumBFS/Yao.jl/issues/176

Comment (GiggleLiu) [reply test/operations.jl]: Also, avoiding comparing `reg` and `reg'` directly is nessesary.

File: src/operations.jl [hunk 2]
```diff
@@ -25,36 +25,37 @@ end
...
-    @eval function Base.$op(lhs::ArrayRegOrAdjointArrayReg{B}, rhs::ArrayRegOrAdjointArrayReg{B}) where B
-        ($op)(state(lhs), state(rhs))
+    for AT in [:ArrayReg, :AdjointArrayReg]
+        @eval function Base.$op(lhs::$AT, rhs::$AT)
+            ($op)(state(lhs), state(rhs))
+        end
     end
 end
 
-Base.:*(op::AbstractMatrix, r::ArrayRegOrAdjointArrayReg) = op * state(r)
```
Comment (Roger-luo) [inline src/operations.jl]: This is something exists a long time. If you want to delete it, you need to add deprecations.

Comment (GiggleLiu) [reply src/operations.jl]: = =,
ok, that's fair.

**Pull request merged**